### ISSUE TITLE
Add stochastic softmax tricks

### ISSUE THREAD

**Issue by gdalle** [axelparmentier/InferOpt.jl#48]
https://arxiv.org/abs/2006.08063

**Comment by gdalle** [comment]
Currently reading their paper. The two main ideas are
1) representing a probability distribution on a polytope as the pushforward through an argmax of a simpler probability distribution on a cost
2) relaxing the argmax into a softmax to make it differentiable

IMO the main difference from InferOpt is that from the start they want to differentiate $\int \text{oracle}(U) p(U | \theta) dU$, whereas we want to differentiate $\text{oracle}(\theta)$ and only use perturbation $U \sim p(\cdot | \theta)$ as an approximation.

Their approach relies on being able to differentiate through a regularized optimizer, which as they point out is not easy. In their supplementary material they suggest either unrolling, a "custom backward pass" (not sure what that is) or finite differences. Implicit differentiation of Frank-Wolfe is another possible method.

Basically, I think once #80 is merged, we'll be able to reimplement all of the specific examples they list in their paper

**Comment by gdalle** [comment]
So the question we need to ask here is: in our setting, if we assume we are able to differentiate through a regularized optimizer, why should we add a perturbation in front of it?
